### PR TITLE
Map view and scaling refactor

### PR DESCRIPTION
**Pull request by christophertubbs**
Refactored scoring to aggregate scores of metrics rather than scores of thresholds and finished the map view that can show results of evaluations in real time.

### CODE REVIEW

File: python/services/evaluationservice/dmod/evaluationservice/test_config.json
Comment (aaraney) [inline python/services/evaluationservice/dmod/evaluationservice/test_config.json:35]: probably want to obfuscate this and all other `address` values.

**Review by aaraney — CHANGES_REQUESTED** [review]
Nice stuff! I left a few comments that should be fairly straightforward to address. However, there is a small amount of data obfuscation that needs to occur before I can approve this.

File: .gitignore
Comment (christophertubbs) [reply .gitignore:116]: It should. Fix coming up.

File: python/lib/core/dmod/core/common/helper_functions.py
Comment (christophertubbs) [reply python/lib/core/dmod/core/common/helper_functions.py:13]: I'm not sure a bound of `type` would work here. It'll work for the `get_subclasses` function (that expects some form of `type`), but the `foreach` function down below it expects it like a C++ template type (function takes an instance of type `_CLASS_TYPE` and calls it on a list of instances of type `_CLASS_TYPE`). That `foreach` should be type unbounded.

Comment (christophertubbs) [reply python/lib/core/dmod/core/common/helper_functions.py:13]: Or I can introduce another `typevar`?

File: python/lib/core/dmod/core/common/helper_functions.py
Comment (christophertubbs) [reply python/lib/core/dmod/core/common/helper_functions.py:269]: I take it you mean something along the lines of:

```python
    concrete_classes = list()

    for subclass in base.__subclasses__():
        if inspect.isabstract(subclass):
            concrete_classes.extend([
                cls
                for cls in get_subclasses(subclass)
                if cls not in concrete_classes
                   and not inspect.isabstract(cls)
            ])
        else:
            concrete_classes.append(subclass)

    return concrete_classes
```

The only reason I avoided it was because it's easier to stick your nose in it and investigate when debugging if the list to iterate is defined earlier. Due to lists of subclasses generally end up pretty short, I avoided the optimization of moving the first order filtering to remove the `if` condition from the for loop to make it as simple as possible.

There's not a huge difference, I'm not super attached to the different process, and the change has already been written above, so I can change it if you'd prefer.

File: python/lib/core/dmod/core/common/helper_functions.py
Comment (aaraney) [reply python/lib/core/dmod/core/common/helper_functions.py:13]: Yeah, that makes sense. I think I made the wrong assumption in reading `_CLASS_TYPE`. I was taking that to mean `type(class) # usually `type` unless another metaclass`, but from this comment I take `_CLASS_TYPE` to mean subtype of `object`. Is that fair? If so, what are your thoughts on `_OBJECT_TYPE` instead?

File: python/lib/core/dmod/core/common/helper_functions.py
Comment (aaraney) [reply python/lib/core/dmod/core/common/helper_functions.py:269]: I'd also rather err on the side of clarity, so your comment is well taken! I was thinking you could just remove the `subclasses` list and replace `concrete_classes` with:

```python
concrete_classes = [
        subclass
        for subclass in base.__subclasses__()
        if not inspect.isabstract(subclass)
    ]
```

Like you mention, its not a big deal at all (the length of `subclasses` is never going to be very long). Ill leave this one up to you. If you want to make the nano-optimization, go for it, I think im just being a little nitpicky.

File: python/lib/core/dmod/core/common/helper_functions.py
Comment (christophertubbs) [reply python/lib/core/dmod/core/common/helper_functions.py:291]: Is this just for naming?

I think it was originally `each`, but I went with `foreach` since that's a more common term used across programming languages. It is a little more obvious than `map` (and `greedy_map` by extension), so I'd prefer some form of term that explicitly states "I'm calling this for everything in the collection" since `for` logic is more of a 100 level idea with `map` being a 200 level.

File: python/lib/core/dmod/core/common/helper_functions.py
Comment (christophertubbs) [reply python/lib/core/dmod/core/common/helper_functions.py:269]: I'm going with your suggestion since that initial `subclasses` collection is just a copy of `base.__subclasses__()` where we just need to iterate.

File: python/lib/core/dmod/core/common/helper_functions.py
Comment (christophertubbs) [reply python/lib/core/dmod/core/common/helper_functions.py:477]: The `places is None` is more of a CYA maneuver than anything else. Callers can technically still pass `truncate_numbers_in_dictionary({"example": 234238.234234232352352}, None)`. The situation will probably never arrive, though. I'm gonna remove it just because it'll just be that rare.

File: python/lib/core/dmod/core/common/helper_functions.py
Comment (christophertubbs) [reply python/lib/core/dmod/core/common/helper_functions.py:480]: It used the comprehension because the input type hint was `typing.Mapping` which doesn't define `copy`, not `dict`. I'm just going to go ahead and change the hinting.

File: python/lib/evaluations/dmod/evaluations/evaluate.py
Comment (christophertubbs) [reply python/lib/evaluations/dmod/evaluations/evaluate.py:376]: Changed to `logging.error`

File: python/lib/evaluations/dmod/evaluations/evaluate.py
Comment (christophertubbs) [reply python/lib/evaluations/dmod/evaluations/evaluate.py:393]: I was just going to use `logging.error` for this, but yeah, raising the `ValueError` is way better.

File: python/lib/evaluations/dmod/evaluations/specification/model.py
Comment (christophertubbs) [reply python/lib/evaluations/dmod/evaluations/specification/model.py:1344]: It's to fulfill the need for the abstract `validate` function from `class Specification(abc.ABC)`. I went ahead and added code to check for validation issues for the internal backend specification.

File: python/lib/evaluations/dmod/evaluations/specification/model.py
Comment (christophertubbs) [reply python/lib/evaluations/dmod/evaluations/specification/model.py:1829]: Yes it was. Good catch.

File: python/lib/metrics/dmod/metrics/communication.py
Comment (christophertubbs) [reply python/lib/metrics/dmod/metrics/communication.py:292]: Controversial approach: I'm leaving the `print` just because that `\n` rendering was a real problem without it, but I'm floating up the message in a new exception as suggested rather than just re-raising.

File: python/lib/metrics/dmod/metrics/scoring.py
Comment (christophertubbs) [reply python/lib/metrics/dmod/metrics/scoring.py:38]: I'm going to need to think about this one since you bring up a really, _really_ good point. This might qualify as another ticket.

File: python/lib/metrics/dmod/metrics/scoring.py
Comment (christophertubbs) [reply python/lib/metrics/dmod/metrics/scoring.py:327]: Pedantic or not, I think it's better phrasing, so I'm updating it.

File: python/lib/metrics/dmod/metrics/scoring.py
Comment (christophertubbs) [reply python/lib/metrics/dmod/metrics/scoring.py:334]: Same response as the above: I dig the new wording.

File: python/services/evaluationservice/dmod/evaluationservice/evaluation_service/choices.py
Comment (christophertubbs) [reply python/services/evaluationservice/dmod/evaluationservice/evaluation_service/choices.py:40]: It's to leverage the formatting of the members to form the list of `(str, str)` tuples that Django expects. `StrEnum` didn't get introduced until python 3.11, so this was the most effective way to ensure that a database friendly string got dynamically tied to a screen friendly string.

File: python/services/evaluationservice/dmod/evaluationservice/evaluation_service/static/evaluation_service/js/active_map.js
Comment (christophertubbs) [reply python/services/evaluationservice/dmod/evaluationservice/evaluation_service/static/evaluation_service/js/active_map.js:26]: I tried to come up with a reason to make that equality loosey-goosey, but couldn't, so I'm changing it.

File: python/services/evaluationservice/dmod/evaluationservice/evaluation_service/static/evaluation_service/js/active_map.js
Comment (christophertubbs) [reply python/services/evaluationservice/dmod/evaluationservice/evaluation_service/static/evaluation_service/js/active_map.js:112]: My IDE doesn't like to parse formatted strings for whatever reason (probably so I have to buy the paid version), so the formatted string version looks real, REAL concerning. I have a couple formatted strings floating about, though, so I'm going to go ahead and slip this one in there too.

File: python/services/evaluationservice/dmod/evaluationservice/test_config.json
Comment (christophertubbs) [reply python/services/evaluationservice/dmod/evaluationservice/test_config.json:35]: That file wasn't even supposed to be there in the first place. Good catch.

File: python/services/evaluationservice/dmod/evaluationservice/evaluation_service/static/evaluation_service/js/active_map.js
Comment (aaraney) [reply python/services/evaluationservice/dmod/evaluationservice/evaluation_service/static/evaluation_service/js/active_map.js]: That's sad. It looks like github's preformat renderer doesn't know how to handle template literals either (see above snippet), at least in Chrome they arent formatted correctly.

File: python/lib/metrics/dmod/metrics/communication.py
Comment (aaraney) [reply python/lib/metrics/dmod/metrics/communication.py:291]: Seems hacky and like something you would probably have to do to make python logging happy...

Comment (aaraney) [reply python/lib/metrics/dmod/metrics/communication.py:291]: Sorry for not reading your inline comment. I dont think I would have commented if I had.

File: python/lib/core/dmod/core/common/helper_functions.py
Comment (aaraney) [reply python/lib/core/dmod/core/common/helper_functions.py:286]: Yeah, just about naming. I was just being pedantic. I don't think _greedy_ was the word I was looking for when I originally commented, instead I think I was looking for _eager_. Either way, I like your logical hierarchy between `for` and `map`. And to be clear, im totally cool leaving this how it is! With that in mind, the only thing I am debating is the function name because it returns a new sequence rather than mutating the existing collection. I am also biased because that is how [`javascipt`'s `foreach` function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach) works.

Connecting that idea back to `map` and `for`, when _I_ think of `map`, I think it will return a new collection that does not modify the original collection (its _pure_; ah yes, the 200 level jargon). Where as `foreach` sounds like it should modify a value in place. Thoughts?

File: .gitignore
Comment (aaraney) [reply .gitignore]: Great, thanks for doing that!

File: python/lib/core/dmod/core/common/helper_functions.py
Comment (aaraney) [reply python/lib/core/dmod/core/common/helper_functions.py]: Great, thanks for the dialog and for compromising!

File: python/lib/evaluations/dmod/evaluations/evaluate.py
Comment (aaraney) [reply python/lib/evaluations/dmod/evaluations/evaluate.py]: Great, thanks!

File: python/lib/metrics/dmod/metrics/scoring.py
Comment (aaraney) [reply python/lib/metrics/dmod/metrics/scoring.py:38]: @christophertubbs, Any movement or decisions made on this front?

File: python/lib/metrics/dmod/metrics/scoring.py
Comment (aaraney) [reply python/lib/metrics/dmod/metrics/scoring.py]: :tada:

File: python/services/evaluationservice/dmod/evaluationservice/evaluation_service/choices.py
Comment (aaraney) [reply python/services/evaluationservice/dmod/evaluationservice/evaluation_service/choices.py:44]: @christophertubbs, can you take a look at this? This will need to be changed before we merge.

**Review by aaraney — COMMENTED** [review]
Sorry, I got caught up in a few meetings this morning and im just now getting around to finishing up this review. All changes you made look great, thanks for making them, however there are still a few prior comments that I would like to talk through and one change that is required before we can merge.

To simplify this process, I am just going comment on this and approve this PR once we've worked through the aforementioned comments / requested changes.

File: python/services/evaluationservice/dmod/evaluationservice/evaluation_service/choices.py
Comment (christophertubbs) [reply python/services/evaluationservice/dmod/evaluationservice/evaluation_service/choices.py:44]: Good catch. I'm going to address this first.

Comment (christophertubbs) [reply python/services/evaluationservice/dmod/evaluationservice/evaluation_service/choices.py]: Alright, I addressed it by using dynamic staticmethods that will format encountered entries into the format that Django expects for Django string choices.

Enums weren't used due to the current lack of support for strings.

File: python/lib/core/dmod/core/common/helper_functions.py
Comment (christophertubbs) [reply python/lib/core/dmod/core/common/helper_functions.py:286]: That's a good point. I updated the function to update the original collection if it was mutable and default to the old behavior if it isn't.

It doesn't have to return anything - that was mostly an afterthought. `map` should be used for that (I'd think), but if someone is calling a function that happens to return a value, why dump the result immediately?

File: python/services/evaluationservice/dmod/evaluationservice/evaluation_service/choices.py
Comment (aaraney) [reply python/services/evaluationservice/dmod/evaluationservice/evaluation_service/choices.py]: Thanks for addressing this! I like the solution you settled on!

**Review by aaraney — APPROVED** [review]
Looks good to me! Thanks for addressing the outstanding comments, @christophertubbs. Merge at will.

File: python/lib/core/dmod/core/common/helper_functions.py
Comment (christophertubbs) [reply python/lib/core/dmod/core/common/helper_functions.py:286]: Post-Mortem update: I'm removing the value storage/returning/updating mechanism and renaming this 'on_each' (since it operates "on each").

If you use the example function `p` from the docstring with this function, it will replace the collection `[1, 2, 3]` with `[None, None, None]` in place since `p` doesn't return anything. I can't think of a good way to detect if the value in the collection _should_ be updated either. You can only tell if the `func` function returns a value if the author annotated the result (not 100% of the time), so you can't use that to indicate that a returned `None` means that there shouldn't be an update.

**Pull request merged**
